### PR TITLE
Implement mutation observer in DOMChildrenOperations

### PR DESCRIPTION
**Pull request by thinkxl**
This pull request (WIP) tries to resolve https://github.com/facebook/react/issues/3218

I have done some basic implementation and tried to add a devtool hook, but have some questions:

* What's the goal of the hooks in `ReactDebugTool.js`, how can I trigger an existing hook to see how it works?
* I'm not sure I'd add tests to this implementation, any suggestion is welcome
* Also, I'd like suggestion on where to add the `.disconnect();` function

### CODE REVIEW

**Comment by thinkxl** [comment]
Working on the tests

File: src/renderers/dom/stack/client/DOMChildrenOperations.js [hunk 1]
```diff
@@ -246,6 +249,29 @@ var DOMChildrenOperations = {
...
+        subtree: true,
+      };
+
+      // eslint-disable-next-line no-inner-declarations
+      function callback(mutations) {
+        warning(false, 'Warning, an outside source is mutating the DOM');
+      }
+
+      var observer = new MutationObserver(callback);
+      var reactRoot = document.querySelector('[data-reactroot]');
```
Comment (edvinerikson) [inline src/renderers/dom/stack/client/DOMChildrenOperations.js]: guess this will break if there is multiple react trees on the page

Comment (thinkxl) [inline src/renderers/dom/stack/client/DOMChildrenOperations.js]: yeah, i should look for all react trees and run the observer on each of them; will fix.

**Comment by thinkxl** [comment]
hey @edvinerikson, could you take a look on the questions above?

**Comment by LeZuse** [comment]
Are you still working on this @thinkxl? I think it would be helpful to write some tests first and get them approved from the React team.

**Comment by syranide** [comment]
Modifying nodes inside empty React leaf DOM-nodes is allowed and a feature. This PR does not consider this and warns regardless. Preferably it should also not rely on `querySelector` but rather put hooks in the correct internal React methods which knows when React roots are mounted and unmounted. As mentioned below, unless there's some deduping, modifications inside nested React roots will emit multiple warnings.

More info https://github.com/facebook/react/issues/3218#issuecomment-232294580

**Comment by thinkxl** [comment]
@syranide thanks for the feedback. I will do the changes to meet the requirements.

**Comment by thinkxl** [comment]
> Modifying nodes inside empty React leaf DOM-nodes

You mean like when a React component inside a container doesn't have children? 

```js
<div id="app">
  <div data-reactroot></div> <!-- empty node, react component -->
</div> <!-- #app -->
```

Or when the tree of React nodes are empty like `<p></p>` or `<p />` and someone wants to add `.innerText` to that?

```js
<div id="app">
  <div data-reactroot>
    <div class="headerr">
      <img src="logo.svg" alt="logo">
    </div> <!-- non-empty node -->
    <p></p> <!-- empty node, user may want to mutate this externally -->
    <div>foo</div> <!-- non-empty node -->
  </div>
</div> <!-- #app -->
```

**Comment by syranide** [comment]
@thinkxl Yeah, anything inside your "non-empty nodes" is free game and mutations should not emit warnings.

@spicyj I'm curious, when it comes to MutationObservers, how do you feel about listeners per React root vs a single global for the document (or per document in the case of iframes)? Having it per React root is a tighter fit but nested React roots will multiply the overhead and nested React roots updating will most likely cause significant overhead due to one or more parent MutationObservers unless you explicitly disconnect parent MutationsObservers but that can also get somewhat expensive. So IMHO a global MutationObserver seems like the simplest and best of these two approaches, but it obviously comes with the downside of being global.

The best option would naturally be to have it per React root but not if it's nested, but if you move a nested React root it might not have a MutationObserver covering it anymore (acceptable?).

**Comment by Hypnosphi** [comment]
Looks like some flag is needed on empty leafs which would somehow (using `Element.closest()` maybe?) block warnings for all potential subtree. Given that, the same flag could be used for nested react trees, which would remove warnings from parent tree, leaving only those from the children.

Another option can be not to use deep observation with `{subtree: true}`, but to attach plain shallow observer to each dom node which needs it.

**Comment by thinkxl** [comment]
> nother option can be not to use deep observation with {subtree: true}, but to attach plain shallow observer to each dom node which needs it.

That might be the way to go, having an observer per node, but, not sure how much that could impact on performance.

**Comment by gaearon** [comment]
As per above comments, this is too strict and will warn for valid cases. I think we might not want to do this after all, as I don't see a way that would be both fast and loose enough. Thanks for trying though!

**Pull request closed without merging**